File: foghttp/_client/process.py
__all__ = ("current_process_id", "forked_process_error")

from dataclasses import dataclass
import os

from ..errors import LifecycleError


@dataclass(slots=True)
class _ProcessState:
    process_id: int


_PROCESS_STATE = _ProcessState(process_id=os.getpid())


def current_process_id() -> int:
    return _PROCESS_STATE.process_id


def forked_process_error(
    *,
    resource: str,
    created_process_id: int,
    current_process_id: int,
) -> LifecycleError:
    return LifecycleError(
        f"FogHTTP {resource} was created in process {created_process_id} "
        f"and cannot be used in forked process {current_process_id}; "
        "create a new client after fork",
    )


def _refresh_current_process_id() -> None:
    _PROCESS_STATE.process_id = os.getpid()


if hasattr(os, "register_at_fork"):
    os.register_at_fork(after_in_child=_refresh_current_process_id)

File: src/py/client/process.rs
use crate::errors::FogHttpLifecycleError;
use pyo3::PyErr;

pub type ProcessId = u32;

pub fn current_process_id() -> ProcessId {
    std::process::id()
}

pub fn client_used_after_fork(
    created_process_id: ProcessId,
    current_process_id: ProcessId,
) -> PyErr {
    resource_used_after_fork("client", created_process_id, current_process_id)
}

pub fn stream_response_used_after_fork(
    created_process_id: ProcessId,
    current_process_id: ProcessId,
) -> PyErr {
    resource_used_after_fork("stream response", created_process_id, current_process_id)
}

fn resource_used_after_fork(
    resource: &str,
    created_process_id: ProcessId,
    current_process_id: ProcessId,
) -> PyErr {
    FogHttpLifecycleError::new_err(format!(
        "FogHTTP {resource} was created in process {created_process_id} and cannot be used in forked process {current_process_id}; create a new client after fork",
    ))
}
